pkg/repository/store.go

```diff
@@ -0,0 +1,242 @@
+package repository
+
+import (
+	"context"
+	"log/slog"
+
+	asyncclient "github.com/metal-stack/metal-apiserver/pkg/async/client"
+	"github.com/metal-stack/metal-apiserver/pkg/db/generic"
+	"github.com/metal-stack/metal-apiserver/pkg/db/metal"
+
+	adminv2 "github.com/metal-stack/api/go/metalstack/admin/v2"
+	apiv2 "github.com/metal-stack/api/go/metalstack/api/v2"
+	ipamv1connect "github.com/metal-stack/go-ipam/api/v1/apiv1connect"
+	mdcv1 "github.com/metal-stack/masterdata-api/api/v1"
+	mdm "github.com/metal-stack/masterdata-api/pkg/client"
+	"github.com/redis/go-redis/v9"
+)
+
+type (
+	Store struct {
+		log   *slog.Logger
+		ds    generic.Datastore
+		mdc   mdm.Client
+		ipam  ipamv1connect.IpamServiceClient
+		async *asyncclient.Client
+	}
+
+	store[S any, E Entity, M Message, C CreateMessage, U UpdateMessage, Q Query] struct {
+		s     *Store
+		typed S
+		repository[E, M, C, U, Q]
+	}
+)
+
+func New(log *slog.Logger, mdc mdm.Client, ds generic.Datastore, ipam ipamv1connect.IpamServiceClient, redis *redis.Client) (*Store, error) {
+	r := &Store{
+		log:   log,
+		mdc:   mdc,
+		ipam:  ipam,
+		ds:    ds,
+		async: asyncclient.New(log, redis),
+	}
+
+	return r, nil
+}
+
+func (r *Store) IP(project string) IP {
+	return r.ip(&ProjectScope{
+		projectID: project,
+	})
+}
+
+func (r *Store) UnscopedIP() IP {
+	return r.ip(nil)
+}
+
+func (r *Store) ip(scope *ProjectScope) IP {
+	return &store[*ipRepository, *metal.IP, *apiv2.IP, *apiv2.IPServiceCreateRequest, *apiv2.IPServiceUpdateRequest, *apiv2.IPQuery]{
+		repository: &ipRepository{
+			r:     r,
+			scope: scope,
+		},
+	}
+}
+
+func (r *Store) Image() Image {
+	repository := &imageRepository{
+		r: r,
+	}
+
+	return &store[*imageRepository, *metal.Image, *apiv2.Image, *adminv2.ImageServiceCreateRequest, *adminv2.ImageServiceUpdateRequest, *apiv2.ImageQuery]{
+		repository: repository,
+		s:          r,
+		typed:      repository,
+	}
+}
+
+func (r *Store) Network(project string) Network {
+	return r.network(&ProjectScope{
+		projectID: project,
+	})
+}
+
+func (r *Store) UnscopedNetwork() Network {
+	return r.network(nil)
+}
+
+func (r *Store) network(scope *ProjectScope) Network {
+	repository := &networkRepository{
+		r:     r,
+		scope: scope,
+	}
+
+	return &store[*networkRepository, *metal.Network, *apiv2.Network, *apiv2.NetworkServiceCreateRequest, *apiv2.NetworkServiceUpdateRequest, *apiv2.NetworkServiceListRequest]{
+		repository: repository,
+		s:          r,
```

Comment:
this is kinda confusing, repository already contains r ?

Reply:
The variable was actually unnecessary anyway. I was able to remove it.